DistroLauncher/Ubuntu/InitTasks.cpp

```diff
@@ -0,0 +1,190 @@
+
+  DWORD exitCode = -1;
+  if (auto hr = api.WslLaunchInteractive(L"which -s cloud-init", FALSE, &exitCode); FAILED(hr)) {
+    Helpers::PrintErrorMessage(hr);
+    return false;
+  }
+  if (exitCode != 0) {
+    _putws(L"Distro version doesn't support initialization tasks.\n");
+    return false;
+  }
+
+  exitCode = -1;
+  if (auto hr = api.WslLaunchInteractive(L"cloud-init status --wait", FALSE, &exitCode);
+      FAILED(hr)) {
+    Helpers::PrintErrorMessage(hr);
+    return false;
+  }
+  // 0 for "success" or 2 for "recoverable error" - sometimes we get that while status shows "done".
+  // https://cloudinit.readthedocs.io/en/latest/explanation/failure_states.html#cloud-init-error-codes
+  // If that's the case we should inspect the system to check whether we should skip or proceed with
+  // user creation.
+  if (exitCode != 0 && exitCode != 2) {
+    wprintf(L"Distro initialization failed with exit code: %u\n", exitCode);
+    return false;
+  }
+
+  if (!checkDefaultUser) {
+    return true;
+  }
+
+  auto uid = defaultUserInWslConf();
+  if (uid == 0) {
+    return false;
+  }
+
+  // Use the same flags as the upstream `DistributionInfo::CreateUser()` function.
+  if (auto hr = api.WslConfigureDistribution(uid, WSL_DISTRIBUTION_FLAGS_DEFAULT); FAILED(hr)) {
+    Helpers::PrintErrorMessage(hr);
+    return false;
+  }
+
+  return true;
+}
+
+namespace {
+namespace fs = std::filesystem;
+std::pair<std::string_view, std::string_view> matchKeyValuePair(std::string_view line);
+std::wstring str2wide(const std::string& str, UINT codePage = CP_THREAD_ACP);
+
+class IniReader {
+ public:
+  explicit IniReader(fs::path path) : file{path} {}
+  // Returns the value of [section].key as a string.
+  // To ease distinguishing the parameters, section is expected to contain the
+  // square-brackets.
+  std::string operator()(std::string_view section, std::string_view key) {
+    if (!seekSection(section)) {
+      return {};
+    }
+    return findValueInCurrentSection(key);
+  }
+  ~IniReader() = default;
+
+ private:
+  // Positions the file stream past the line where section is declared.
+  // Returns false if section is not found.
+  // Note: this function always starts from the beginning of the file.
+  bool seekSection(std::string_view section) {
+    std::string line;
+    file.seekg(0);
+    while (std::getline(file, line) && line.find(section, 0) == std::string::npos) {}
+    return !file.eof();
+  }
+
+  // Iterates file until key, another section or EOF is found.
+  // If the key is found, its value is returned. Empty string, otherwise.
+  std::string findValueInCurrentSection(std::string_view key) {
+    for (std::string line; std::getline(file, line);) {
+      if (line.find('[', 0) != std::string::npos) {
+        // found another section, so value not found.
+        return {};
+      }
+      if (auto [k, v] = matchKeyValuePair(line); k == key) {
+        return std::string{v};
+      }
+    }
+
+    return {};
+  }
```

Comment:
This stops on the first occurence of `[section].key`. For a moment I considered that could be a mistake, but indeed that's how WSL parses ini:

![image](https://github.com/ubuntu/WSL/assets/11138291/371d75b5-5231-468b-b21a-69bdb12add40)

I'll add one comment about this for the future.